src/hotspot/share/opto/vector.cpp

```diff
@@ -336,30 +347,87 @@ Node* PhaseVector::expand_vbox_node_helper(Node* vbox,
+      Node* new_box = expand_vbox_node_helper(vec_box, vbox->in(i), vect, box_type, vect_type);
       new_phi->set_req(i, new_box);
     }
     new_phi = C->initial_gvn()->transform(new_phi);
     return new_phi;
   } else if (vbox->is_Proj() && vbox->in(0)->Opcode() == Op_VectorBoxAllocate) {
     VectorBoxAllocateNode* vbox_alloc = static_cast<VectorBoxAllocateNode*>(vbox->in(0));
-    return expand_vbox_alloc_node(vbox_alloc, vect, box_type, vect_type);
+    return expand_vbox_alloc_node(vec_box, vbox_alloc, vect, box_type, vect_type);
   } else {
     assert(!vbox->is_Phi(), "");
     // TODO: assert that expanded vbox is initialized with the same value (vect).
     return vbox; // already expanded
   }
 }
 
-Node* PhaseVector::expand_vbox_alloc_node(VectorBoxAllocateNode* vbox_alloc,
+Node* PhaseVector::expand_vbox_alloc_node_mf(Node* vbox,
+                                             VectorBoxAllocateNode* vbox_alloc,
+                                             Node* value,
+                                             const TypeInstPtr* box_type,
+                                             const TypeVect* vect_type) {
+  assert(vbox->isa_InlineType(), "");
+  JVMState* jvms = clone_jvms(C, vbox_alloc);
+  GraphKit kit(jvms);
+  PhaseGVN& gvn = kit.gvn();
+
+  ciInstanceKlass* box_klass = box_type->instance_klass();
+  BasicType bt = vect_type->element_basic_type();
+  int num_elem = vect_type->length();
+  int elem_size = type2aelembytes(bt);
+
+  const TypeKlassPtr* klass_type = box_type->as_klass_type();
+  Node* klass_node = kit.makecon(klass_type);
+  Node* buffer_mem = kit.new_instance(klass_node, NULL, NULL, /* deoptimize_on_exception */ true, vbox->as_InlineType());
+  // TODO: Re-use existing value storage routine from InlineTypeNode.
+  //vbox->as_InlineType()->store(&kit, buffer, buffer, box_klass);
+
+  // Store the vector value into the array.
+  // (The store should be captured by InitializeNode and turned into initialized store later.)
+  ciSymbol* payload_sig = ciSymbol::make(VectorSupport::get_vector_payload_field_signature(bt, num_elem)->as_C_string());
+  ciSymbol* payload_name = ciSymbol::make(vmSymbols::payload_name()->as_C_string());
+  ciField* payload = box_klass->get_field_by_name(payload_name, payload_sig, false);
+
+  Node* buffer_start_adr = kit.basic_plus_adr(buffer_mem, payload->offset());
+  const TypePtr* buffer_adr_type = buffer_start_adr->bottom_type()->is_ptr();
+  Node* buffer_mem_start = kit.memory(buffer_start_adr);
+  Node* vstore = gvn.transform(StoreVectorNode::make(0,
+                                                     kit.control(),
+                                                     buffer_mem_start,
+                                                     buffer_start_adr,
+                                                     buffer_adr_type,
+                                                     value,
+                                                     num_elem));
```

Comment:
Can we simply call `InlineTypeNode::buffer()` here?

Reply:
Logic is now simplified, vector box IR also comply with the composition of concrete vector class which is a value type encapsulating primitive type payload.
![image](https://user-images.githubusercontent.com/59989778/230422964-80eb40cc-0162-4de2-9c61-e5a42495fd1e.png)

This allows direct usage of InlineTypeNode::store routine which does a recursive field traversal to store scalarized fields into buffer.